src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Cors.Core;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.WebUtilities;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// A filter which applies the given <see cref="CorsPolicy"/> and adds appropriate response headers.
+    /// </summary>
+    public class CorsAuthorizationFilter : ICorsAuthorizationFilter
+    {
+        private readonly string _corsPolicyName;
+
+        /// <summary>
+        /// Creates a new instace of <see cref="CorsAuthorizationFilter"/>.
+        /// </summary>
+        /// <param name="policyName">The policy name which needs to be applied.</param>
+        public CorsAuthorizationFilter(string policyName)
+        {
+            _corsPolicyName = policyName;
+        }
+
+        /// <inheritdoc />
+        public async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // If this filter is not closest to the action, it is not applicable.
+            if (!IsClosestToAction(context.Filters))
+            {
+                return;
+            }
+
+            var httpContext = context.HttpContext;
+            var request = httpContext.Request;
+            if (request.Headers.ContainsKey(CorsConstants.Origin))
+            {
+                var corsPolicyProvider = httpContext.RequestServices.GetRequiredService<ICorsPolicyProvider>();
```

Comment:
Another thing that we can potentially do is instead of creating a `CorsAuthorizationFilter ` we create a `CorsAuthorizationFilterFactory : IFilterFactory` which in its `CreateInstance` can create the  `CorsAuthorizationFilter`. 
The only downside is since we require to pass in the policy name to the constructor the factory cannot be DI ed.